### ISSUE TITLE
Bad rendering method in the OpenGLES based demo.

### ISSUE THREAD

**Issue by zwcloud** [LayoutFarm/Typography#45]
From [this][1], we can see each glyph uses a drawcall and it harm performance badly when we are rendering a string containing many characters.

I think we should apply the offset when we fill the vertex buffer. I mean, bake the offset into vertex instead of offset it in vertex shader, just like [what I have done before][2] in my original OpenGLES demo.

[1]:https://github.com/LayoutFarm/Typography/blob/bb3bfa05aedaf958446c9f1c75c1b9908e2a9b13/Demo/Android/Xamarin.OpenGL/App/SimpleCanvas.cs#L118]
[2]:https://github.com/zwcloud/Typography/blob/android_demo/Demo/Android/Xamarin.OpenGL/TextRenderer/TextPrinter/GlyphPathBuilder.cs#L107

**Comment by prepare** [comment]
Do you have the comparison results of method A and B ?
please let me know

**Comment by zwcloud** [comment]
> Do you have the comparison results of method 1 and 2 ?
No. But I think you may have a look at [this SO post](http://stackoverflow.com/questions/4853856/why-are-draw-calls-expensive). I should say it is common sense to reduce drawcalls.

**Comment by prepare** [comment]
at this time,
I think the performance problem may come from ...
I load triangles mesh into GPU every time for every glyph.

---
solutions:
1.  use VBO, I will implement this later, see https://github.com/prepare/PixelFarm-dev/issues/38
2.  use Glyph texture, for general font. (use Triangles for very large font)
    see https://github.com/LayoutFarm/Typography/tree/master/Typography.MsdfGen
    

this is sample of msdf texture font, fast and scalable!
![msdf_test_03](https://cloud.githubusercontent.com/assets/7447159/25564423/3686e0dc-2ddd-11e7-97f5-c34ba2d21762.png)
_pic 1: msdf texture font sample_

![subpix06](https://cloud.githubusercontent.com/assets/7447159/24132413/017536fc-0e28-11e7-953a-989c3acd1f4c.png)
_pic 2: subpixel font texture with GLES2_


latest subpixel font texture rendering with GLES2 (PREVIEW) gives a very cripes font!
![glyph_analysis_e_26](https://cloud.githubusercontent.com/assets/7447159/25780573/b9d2172e-3353-11e7-8558-ccb03855c56f.png)

_pic 3: Tahoma, 8 pt, (1) No AutoFit , (2) with AutoFit, 
Tahoma, 10pt, (3) No AutoFit, (4) with AutoFit_

**Comment by zwcloud** [comment]
> load triangles mesh into GPU every time for every glyph

Yes. That should be avoided.

MSDF_ is a good way to render glyphs on any platforms. But personally, I prefer 3D-graphics-only method to render glyph. _ Because pixels in a **mobile screen** is so small that anti-aliasing is no longer needed and it's quite easy to implement (just fill the line-segments and bezier curves). We may possibly find some way to improve the rendered result by adjusting it in the vertex or pixel shader in the future for desktop platforms.

My opinion may not be right. It's just my personal preference to try something new. 😄

**Comment by prepare** [comment]
Now, I'm tackle this with VBO.
